### PR TITLE
Use runtime instead of threading trace through evaluator

### PR DESCRIPTION
**Pull request by jack-williams**
Candidate approach of trying to tidy up some of the handling of pending data.

### CODE REVIEW

File: packages/core/nano/src/core.ts [hunk 1]
```diff
@@ -1,21 +1,26 @@
...
+    AsPrimitive = "value",
+}
+
 // TODO: Support Completions
 export interface CalcObj<O> {
-    read(property: string, origin: O, ...args: any[]): CalcValue<O> | Pending<CalcValue<O>>;
+    read: (property: string, origin: O, ...args: any[]) => CalcValue<O> | Pending<CalcValue<O>>;
 }
 
-export interface CalcFun<O = unknown> {
```
Comment (jack-williams) [inline packages/core/nano/src/core.ts]: Not convinced on the merit of removing this default

File: packages/core/nano/src/core.ts [hunk 2]
```diff
@@ -71,47 +79,57 @@ export function makeTracer(): [Pending<unknown>[], Trace] {
...
+    ifS: <A>(cond: Delayed<boolean>, cont: (cond: boolean) => Delayed<A>) => Delayed<A>;
+    app1: <O, A, B>(origin: O, op: (runtime: Runtime, origin: O, expr: A) => B, expr: Delayed<A>) => Delayed<B>;
+    app2: <O, A, B, C>(origin: O, op: (runtime: Runtime, origin: O, l: A, r: B) => C, l: Delayed<A>, r: Delayed<B>) => Delayed<C>;
+    appN: <O>(origin: O, fn: Delayed<CalcValue<O>>, args: Delayed<CalcValue<O>>[]) => Delayed<CalcValue<O>>;
 }
 
-function read<O>(trace: Trace, host: O, context: Delayed<CalcValue<O>>, prop: string): Delayed<CalcValue<O>> {
-    return isDelayed(context) ? delay : typeof context === "object" ? trace(context.read(prop, host)) : readOnNonObjectError;
-}
+class CoreRuntime {
```
Comment (jack-williams) [inline packages/core/nano/src/core.ts]: Only reason for using a class here is reduction memory footprint of creating closures for each call of a formula

File: packages/core/nano/src/core.ts [hunk 3]
```diff
@@ -173,5 +191,9 @@ export const ops: OpContext = {
     negate: liftUnaryOp((x: any) => -x),
 };
 
-export type Formula = <O>(host: O, context: CalcValue<O>) => [Pending<unknown>[], Delayed<CalcValue<O>>];
+export type Formula = <O>(origin: O, context: CalcObj<O>) => [Pending<unknown>[], Delayed<CalcValue<O>>];
```
Comment (jack-williams) [inline packages/core/nano/src/core.ts]: I couldn't think of a good reason for having a non-object as the context for a formula, so I changed this.

File: packages/core/nano/src/core.ts (hunk 2 above)
Comment (DLehenbauer) [reply packages/core/nano/src/core.ts]: I like it.

**Pull request merged**